### PR TITLE
ISPN-11541 Add internal metadata to write commands

### PR DESCRIPTION
**Pull request by pruivo**
https://issues.redhat.com/browse/ISPN-11541

### CODE REVIEW

**Comment by tristantarrant** [comment]
Should we perfack this ?

**Comment by ryanemerson** [comment]
@pruivo Is this PR still valid in light of https://github.com/infinispan/infinispan/pull/7858#issuecomment-607286661?

```quote
Probably, it is best just to create a simple class with 2 fields (irac version and opt-tx version)
```

**Comment by pruivo** [comment]
> @pruivo Is this PR still valid in light of [#7858 (comment)](https://github.com/infinispan/infinispan/pull/7858#issuecomment-607286661)?
> 
> ```
> Probably, it is best just to create a simple class with 2 fields (irac version and opt-tx version)
> ```

yes. I have already the `InternalCacheEntry` updated and if I change the class, it would break this PR and create a bunch of conflicts in the next ones. 
I would rather have everything integrated with `MetaParamsInternalMetadata` and change it later in one go.

**Comment by pruivo** [comment]
run performance tests please

**Comment by ryanemerson** [comment]
> I would rather have everything integrated with `MetaParamsInternalMetadata` and change it later in one go.

How far down the line are we talking? This touches a lot of classes, so it makes it somewhat of a moving target for us working in the same area.

**Comment by pruivo** [comment]
The persistence is missing. My plans are to open a PR after this one.
It includes updating `MarshallableEntry` and the `CacheStore` implementations

File: core/src/main/java/org/infinispan/commands/functional/AbstractWriteManyCommand.java [hunk 1]
```diff
@@ -20,6 +24,7 @@
    long flags;
    DataConversion keyDataConversion;
    DataConversion valueDataConversion;
+   Map<Object, MetaParamsInternalMetadata> internalMetadataMap;
```
Comment (wburns) [inline core/src/main/java/org/infinispan/commands/functional/AbstractWriteManyCommand.java:27]: Is there no way we can trim this somehow... a ConcurrentHashMap for every functional command is a lot of bloat. Why not just 1 like we did with the other metadata?

File: core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java [hunk 2]
```diff
@@ -125,9 +124,10 @@ private AsyncInterceptorChain buildInterceptorChain() {
       }
       interceptorChain.appendInterceptor(createInterceptor(new InvocationContextInterceptor(), InvocationContextInterceptor.class), false);
 
-      if (!configuration.transaction().transactionMode().isTransactional()) {
-         interceptorChain.appendInterceptor(createInterceptor(new VersionInterceptor(), VersionInterceptor.class), false);
-      }
+      //TODO! why adding version for replace?
+      //if (!configuration.transaction().transactionMode().isTransactional()) {
+      //   interceptorChain.appendInterceptor(createInterceptor(new VersionInterceptor(), VersionInterceptor.class), false);
+      //}
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java:130]: Is this still meant to be commented?

If we're removing this, should it be a separate commit/Jira?

File: core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java [hunk 3]
```diff
@@ -733,6 +733,7 @@ protected IntSet getOwnedSegments(ConsistentHash consistentHash) {
       InternalMetadataImpl metadata = new InternalMetadataImpl(e);
       PutKeyValueCommand put = commandsFactory.buildPutKeyValueCommand(e.getKey(), e.getValue(), segmentId,
                                                                        metadata, STATE_TRANSFER_FLAGS);
+      put.setInternalMetadata(e.getInternalMetadata());
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java:736]: We should add `MetaParamsInternalMetadata` to `CommandsFactory#buildPutKeyValueCommand`.

Similarly, I noticed that the internal metadata is not  set in `ProtobufMetadataManagerInterceptor`. How are internal commands handled by xsite, should this be set there as well?

File: core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java [hunk 4]
```diff
@@ -208,6 +207,10 @@ private XSiteStatePushCommand newStatePushCommand(List<XSiteState> stateList) {
       return commandsFactory.buildXSiteStatePushCommand(stateList.toArray(new XSiteState[0]), 0);
    }
 
+   private ClusteringDependentLogic getClusteringDependentLogic() {
+      return cache.getComponentRegistry().getComponent(ClusteringDependentLogic.class);
+   }
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java:212]: Unless this method will be used more in follow up changes, this method seems unnecessary.

**Comment by ryanemerson** [comment]
> The persistence is missing. My plans are to open a PR after this one.
> It includes updating `MarshallableEntry` and the `CacheStore` implementations

I have looked through the code and for the most part it seems fine. However it is not easy reviewing knowing that nearly all of the changes in this PR are going to change. Let's get the the follow ups in quickly (I'll make sure I'm timely with a review). I'll hold off rebasing my marshalling branch until then and progress locally, as it's too much of a moving target otherwise due to all of the areas this touches. Plus a simple class containing the two versions will make my life simpler in the long run.

File: core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java (hunk 2 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java:130]: I forgot to ask @gustavonalle about it. Is the VersionInterceptor still needed? 
It sets some version for the replace commands and I didn't find any usage of that version. 
It breaks the xsite tests if non-tx cache (on site 1) backups to a tx cache (on site 2).

File: core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java (hunk 3 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java:736]: > We should add `MetaParamsInternalMetadata` to `CommandsFactory#buildPutKeyValueCommand`.
> 
I didn't add it because I want to try to avoid touching too much interface... it would require a lot of changes everywhere and most of the places, the value is `null`. 

> Similarly, I noticed that the internal metadata is not set in `ProtobufMetadataManagerInterceptor`.
Is it needed? and set it to what exactly? (sorry, no idea what ProtobufMetadataManagerInterceptor is)

The current xsite impl doesn't use it. 
In the new one, the internal metadata is set by the primary owner before replicating it to the backup owners. 
It is also used by the state transfer so the joiners can set the correct versions.

File: core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java (hunk 4 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java:212]: I can remove it. It isn't used yet...

File: core/src/main/java/org/infinispan/commands/functional/AbstractWriteManyCommand.java (hunk 1 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/commands/functional/AbstractWriteManyCommand.java:27]: The internal metadata is stored per key. Each key will have a different version.

File: core/src/main/java/org/infinispan/commands/functional/AbstractWriteManyCommand.java [hunk 5]
```diff
@@ -30,13 +35,15 @@ protected AbstractWriteManyCommand(CommandInvocationId commandInvocationId,
       this.flags = params.toFlagsBitSet();
       this.keyDataConversion = keyDataConversion;
       this.valueDataConversion = valueDataConversion;
+      this.internalMetadataMap = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/commands/functional/AbstractWriteManyCommand.java:38]: no, otherwise I will get an `UnsupportedException` when I try to put metadata there.
It can be `null` but then I will have to put some `null` checks before get/put and do a proper synchronization around it. I would rather not do it.

**Comment by pruivo** [comment]
@ryanemerson updated!
thanks for the review.

File: core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java (hunk 2 above)
Comment (gustavonalle) [reply core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java:130]: I am not sure TBH. This is a relic from the compat mode times

**Comment by ryanemerson** [comment]
@diegolovison Perk ack didn't run for some reason. Can you take a look?

File: core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java (hunk 2 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java:130]: oh... I found a failure... EmbeddedRestMemcachedHotRodTest fails without it.
I need to drop the version in the receiver site... bah...

**Comment by diegolovison** [comment]
run performance tests please

**Comment by diegolovison** [comment]
run performance tests please

**Comment by pruivo** [comment]
perf looks good @ryanemerson

File: core/src/main/java/org/infinispan/remoting/responses/PrepareResponse.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+   }
+
+   @Override
+   public boolean isSuccessful() {
+      return true;
+   }
+
+   @Override
+   public boolean isValid() {
+      return true;
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/remoting/responses/PrepareResponse.java:57]: Unnecessary override of `isValid`, it's the same as `ValidResponse` impl.

File: core/src/main/java/org/infinispan/remoting/responses/PrepareResponse.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+   }
+
+   @Override
+   public Object getResponseValue() {
+      return null; //not used!
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/remoting/responses/PrepareResponse.java:62]: Should we `throw new UnsupportedOperationException();` here the same as in `UnsureResponse` as it should never be called?

File: core/src/main/java/org/infinispan/remoting/responses/PrepareResponse.java (hunk 6 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/remoting/responses/PrepareResponse.java:57]: :+1:

File: core/src/main/java/org/infinispan/remoting/responses/PrepareResponse.java (hunk 7 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/remoting/responses/PrepareResponse.java:62]: I can change it
:+1:

**Comment by ryanemerson** [comment]
Thanks @pruivo
